File: solutions/haskell/163-longest-repeating-character-replacement.hs
{-
  Problem 163: Longest Repeating Character Replacement (LeetCode 424)
  Difficulty: Med
  Language: Haskell
-}
import qualified Data.Map.Strict as Map

characterReplacement :: String -> Int -> Int
characterReplacement s k =
  let go _ _ _ best [] = best
      go left maxC freq best (c:rest) =
        let freq' = Map.insertWith (+) c 1 freq
            mc = max maxC (freq' Map.! c)
            right = length s - length rest
            wLen = right - left
        in if wLen - mc > k
           then let lc = s !! left
                    freq'' = Map.adjust (subtract 1) lc freq'
                in go (left+1) mc freq'' (max best (wLen - 1)) rest
           else go left mc freq' (max best wLen) rest
  in go 0 0 Map.empty 0 s

File: solutions/scala/163-longest-repeating-character-replacement.scala
/**
 * Problem 163: Longest Repeating Character Replacement (LeetCode 424)
 * Difficulty: Med
 * Language: Scala
 */
def characterReplacement(s: String, k: Int): Int = {
  s.indices.foldLeft((0, 0, Map.empty[Char, Int], 0)) {
    case ((left, maxC, freq, best), right) =>
      val ch = s(right)
      val f = freq.updated(ch, freq.getOrElse(ch, 0) + 1)
      val mc = math.max(maxC, f(ch))
      if (right - left + 1 - mc > k) {
        val lf = f.updated(s(left), f(s(left)) - 1)
        (left + 1, mc, lf, math.max(best, right - left))
      } else (left, mc, f, math.max(best, right - left + 1))
  }._4
}
